superduperdb/cluster/task_workflow.py

```diff
@@ -1,66 +1,73 @@
-import datetime
 import networkx
+import copy
+import datetime
+import dataclasses as dc
+import functools
+import typing as t
 import uuid
 
 from superduperdb.cluster.function_job import function_job
 from superduperdb.cluster.job_submission import dask_client
 
+Self = t.Any
+
 
+@dc.dataclass
 class TaskWorkflow:
-    def __init__(self, database):
-        self.G = networkx.DiGraph()
-        self._path_lengths = None
-        self.database = database
+    database: t.Any
+    G: networkx.DiGraph = dc.field(default_factory=networkx.DiGraph)
 
-    def add_edge(self, node1, node2):
+    def add_edge(self, node1, node2) -> None:
         self.G.add_edge(node1, node2)
 
-    def add_node(self, node, data=None):
+    def add_node(self, node, data=None) -> None:
         data = data or {}
         self.G.add_node(node)
-        for k, v in data.items():
-            self.G.nodes[node][k] = v
+        self.G.nodes[node].update(data)
 
-    def __mul__(self, other):
+    def asjson(self) -> t.Dict:
+        return networkx.adjacency_data(self.G)
+
+    def __imul__(self, other: Self) -> Self:
         for node in other.nodes:
             self.G.add_node(node)
         for edge in other.edges:
             self.G.add_edge(*edge)
         return self
 
-    def __add__(self, other):
+    def __iadd__(self, other: Self) -> Self:
```

Comment:
```suggestion
    def __iadd__(self, other: Self) -> 'TaskWorkflow':
```
Only with python 3.11 and [PEP-673](https://peps.python.org/pep-0673/) we get possibility to use `self` as an annotation. Until the minimum supported version is 3.11 or above, it is preferable to consider the suggestion